client_test.go

```diff
@@ -1813,3 +1813,112 @@ func TestClientTLSHandshakeTimeout(t *testing.T) {
 		t.Errorf("resulting error not a timeout: %v\nType %T: %#v", err, err, err)
 	}
 }
+
+func TestClient_WaitPrefer(t *testing.T) {
+	s := startEchoServer(t, "tcp", "127.0.0.1:")
+	defer s.Stop()
+	addr := "http://" + s.Addr()
+	var (
+		limitClient = &Client{
+			MaxConnsPerHost: 2,
+		}
+		waitClient = &Client{
+			PreferWaitConn:  true,
+			MaxConnsPerHost: 2,
+			ReadTimeout:     time.Minute,
+		}
+		wg             sync.WaitGroup
+		noFreeConnErrs int32
+	)
+
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for j := 0; j < 10; j++ {
+				_, _, err := limitClient.Get(nil, addr)
+				if err != nil {
+					if err == ErrNoFreeConns {
+						atomic.AddInt32(&noFreeConnErrs, 1)
+					} else {
+						t.Fatal("unexpected error:", err)
+					}
+				}
+			}
+		}()
+	}
+	wg.Wait()
+	if noFreeConnErrs == 0 {
+		t.Fatal("should be some ErrNoFreeConns")
+	}
+	t.Log("noFreeConnErrs:", noFreeConnErrs)
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			testClientGet(t, waitClient, addr, 30)
+			testClientPost(t, waitClient, addr, 10)
+		}()
+	}
+	wg.Wait()
+}
+
+// TestClient_DoTimeout_PreferWaitConn  running concurrent 5 request  with 1  wait prefer connection,server blocking 300ms.
+// below is simulate activity.
+// id\timeline | start  | 3000ms+  | 6000ms+  | 9000ms+   | 10000ms         |  12000ms(background)|
+// 1            running     OK
+// 2            wait     running     OK
+// 3            wait     wait       running    OK
+// 4            wait     wait       wait       wait        timeout(running) |  (ok)               |
+// 5            wait     wait       wait       running     timeout(wait)    |  (cancel)           |
+//
+// on 0,3000,6000,9000 ms will sending request to server.
+// request at 9000ms will timeout because did not get response in time.
+// one request wait conn all the time  because of no free conns then cancel since timeout.
+
+func TestClient_DoTimeout_PreferWaitConn(t *testing.T) {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer listener.Close()
+	addr := "http://" + listener.Addr().String()
+	var (
+		serverGot int32
+		timeout   int32
+		wg        sync.WaitGroup
+	)
+	s := &Server{
+		Handler: func(ctx *RequestCtx) {
+			atomic.AddInt32(&serverGot, 1)
+			time.Sleep(time.Second * 3)
```

Comment:
Nice test. But can it also work with a bit shorter times? Maybe tens of milliseconds instead of seconds. That way our test suite doesn't take that long.

Reply:
ci should this bag

Reply:
I don't understand? People making changes to fasthttp will run `go test` locally and having to wait an extra 3 seconds for this is very annoying. The test should work just as fine if this was `time.Millisecond * 30` and below `time.Millisecond * 100`.